x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -301,28 +310,77 @@ && canAllocate(shard, allocation) == false) {
             }
 
             // track these to ensure we do not double account if they both cannot remain and allocated due to storage.
-            Set<ShardRouting> unmovableShards = candidates.stream()
+            List<ShardNodeDecisions> unmovableShardNodeDecisions = candidates.stream()
                 .filter(s -> allocatedToTier(s, allocation))
-                .filter(s -> cannotRemainDueToStorage(s, allocation))
-                .collect(Collectors.toSet());
+                .flatMap(shard -> cannotRemainDueToStorage(shard, allocation).stream())
+                .toList();
+            List<ShardNodeDecisions> otherNodesOnTierAllocationDecisions = unmovableShardNodeDecisions.stream()
+                .map(ShardNodeDecisions::shard)
+                .map(
+                    shard -> new ShardNodeDecisions(
+                        shard,
+                        state.getRoutingNodes()
+                            .stream()
+                            .filter(n -> nodeTierPredicate.test(n.node()))
+                            .filter(n -> shard.currentNodeId().equals(n.nodeId()) == false)
+                            .map(n -> new NodeDecision(n.node(), allocationDeciders.canAllocate(shard, n, allocation)))
+                            .toList()
+                    )
+                )
+                .toList();
+            Set<ShardRouting> unmovableShards = unmovableShardNodeDecisions.stream().map(e -> e.shard).collect(Collectors.toSet());
+
             long unmovableBytes = unmovableShards.stream()
                 .collect(Collectors.groupingBy(ShardRouting::currentNodeId))
                 .entrySet()
                 .stream()
                 .mapToLong(e -> unmovableSize(e.getKey(), e.getValue()))
                 .sum();
 
-            List<ShardRouting> unallocatedShards = candidates.stream()
+            List<ShardNodeDecisions> unallocatedShardNodeDecisions = candidates.stream()
                 .filter(Predicate.not(unmovableShards::contains))
-                .filter(s1 -> cannotAllocateDueToStorage(s1, allocation))
+                .flatMap(shard -> cannotAllocateDueToStorage(shard, allocation).stream())
                 .toList();
-            long unallocatableBytes = unallocatedShards.stream().mapToLong(this::sizeOf).sum();
+            List<ShardNodeDecisions> canRemainDecisionsForUnallocatedShards = unallocatedShardNodeDecisions.stream()
+                .filter(snd -> snd.nodeDecisions.size() > 0)
+                .map(ShardNodeDecisions::shard)
+                .map(
+                    shard -> new ShardNodeDecisions(
+                        shard,
+                        List.of(
+                            new NodeDecision(
+                                allocation.routingNodes().node(shard.currentNodeId()).node(),
+                                allocationDeciders.canRemain(shard, allocation.routingNodes().node(shard.currentNodeId()), allocation)
```

Comment:
I think we need to enable debug decisions when calculating these. Same for the `canAllocate` part above.

Also, we should add a test that verifies that we have debug information included.

Reply:
Done! I've added a helper method `withAllocationDebugEnabled`, so we can turn the debug decision on and off and keep the calculation in the short lambda form.